std/digest/xxh.d

```diff
@@ -0,0 +1,3177 @@
+/**
+ * Computes xxHash hashes of arbitrary data. xxHash hashes are either uint32_t,
+ * uint64_t or uint128_t quantities that are like a
+ * checksum or CRC, but are more robust and very performant.
+ *
+$(SCRIPT inhibitQuickIndex = 1;)
+
+$(DIVC quickindex,
+$(BOOKTABLE ,
+$(TR $(TH Category) $(TH Functions)
+)
+$(TR $(TDNW Template API) $(TD $(MYREF XXHTemplate)
+)
+)
+$(TR $(TDNW OOP API) $(TD $(MYREF XXH32Digest))
+)
+$(TR $(TDNW Helpers) $(TD $(MYREF xxh32Of))
+)
+)
+)
+
+ * This module conforms to the APIs defined in `std.digest`. To understand the
+ * differences between the template and the OOP API, see $(MREF std, digest).
+ *
+ * This module publicly imports $(MREF std, digest) and can be used as a stand-alone
+ * module.
+ *
+ * License:   $(HTTP www.boost.org/LICENSE_1_0.txt, Boost License 1.0).
+ *
+ * CTFE:
+ * Digests do not work in CTFE
+ *
+ * Authors:
+ * Carsten Schlote, Piotr Szturmaj, Kai Nacke, Johannes Pfau $(BR)
+ * The routines and algorithms are provided by the xxhash.[ch] source
+ * provided at $(I git@github.com:Cyan4973/xxHash.git).
+ *
+ * References:
+ *      $(LINK2 https://github.com/Cyan4973/xxHash, GitHub website of project)
+ *
+ * Source: $(PHOBOSSRC std/digest/xxh.d)
+ *
+ */
+
+/* xxh.d - A wrapper for the original C implementation */
+module std.digest.xxh;
+
+version (X86)
+    version = HaveUnalignedLoads;
+else version (X86_64)
+    version = HaveUnalignedLoads;
+
+///
+@safe unittest
+{
+    //Template API
+    import std.digest.md;
+
+    //Feeding data
+    ubyte[1024] data;
+    XXH_32 xxh;
+    xxh.start();
+    xxh.put(data[]);
+    xxh.start(); //Start again
+    xxh.put(data[]);
+    auto hash = xxh.finish();
+}
+
+///
+@safe unittest
+{
+    //OOP API
+    import std.digest.md;
+
+    auto xxh = new XXH32Digest();
+    ubyte[] hash = xxh.digest("abc");
+    assert(toHexString(hash) == "32D153FF", "Got " ~ toHexString(hash));
+
+    //Feeding data
+    ubyte[1024] data;
+    xxh.put(data[]);
+    xxh.reset(); //Start again
+    xxh.put(data[]);
+    hash = xxh.finish();
+}
+
+public import std.digest;
+
+/* --- Port of C sources (release 0.8.1) to D language below ---------------- */
+
+enum XXH_NO_STREAM = false;
+enum XXH_SIZE_OPT = 0;
+enum XXH_FORCE_ALIGN_CHECK = true;
+enum XXH32_ENDJMP = false;
+
+private import core.bitop : rol, bswap;
+
+/* *************************************
+*  Misc
+***************************************/
+
+alias XXH32_hash_t = uint;
+alias XXH64_hash_t = ulong;
+struct XXH128_hash_t
+{
+    XXH64_hash_t low64; /** `value & 0xFFFFFFFFFFFFFFFF` */
+    XXH64_hash_t high64; /** `value >> 64` */
+}
```

Comment:
The code was taken from the original sources. It's the fallback, if there is no __int128_t available for your target sytsem or compiler. However, there is also 128bit code available in the original sources, so we can ad this with version, e.g. version(Have128Bit) or similiar.

Reply:
I tested the use of core.int128.Cent. Instead of inline code it just resulted into a subroutine call to Cent.mul(). The manual multiplication code provided by XXH is inlined and looks much better. So I disabled use of core.int128 for now, although it works and passes the unittests.

Reply:
> I tested the use of core.int128.Cent. Instead of inline code it just resulted into a subroutine call to Cent.mul(). The manual multiplication code provided by XXH is inlined and looks much better. So I disabled use of core.int128 for now, although it works and passes the unittests.

Maybe we should force those to be inlined.